### PR TITLE
fix: Don't recompile contents of DafnySourceAttribute from DLLs

### LINKED ISSUE #466: Dafny source in DafnySourceAttribute from DLLs should not be re-compiled to C#
This issue gets in the way of encapsulating Dafny code with C# externals inside a DLL. Example:

```
// B.dfy
module {:extern "B"} B {
  class {:extern "Foo"} Foo {
    constructor() {}
    method {:extern "Bar"} Bar()
  }
}
```
```
// BImpl.cs
namespace B {
  public partial class Foo {
    public void Bar() {}
  }
}
```
```
// A.dfy
module A { 
  import B
  method Splat() {
    var foo := new B.Foo();
    foo.Bar();
  }
}
```
Compiling everything together at once works:
```
$ dafny A.dfy B.dfy BImpl.cs
Dafny 2.3.0.10506

Dafny program verifier finished with 1 verified, 0 errors
Compiled program written to A.cs
Compiled assembly into A.dll
```
But trying to compile the B module into a separate DLL doesn't:
```
$ dafny B.dfy BImpl.cs
Dafny 2.3.0.10506

Dafny program verifier finished with 0 verified, 0 errors
Compiled program written to B.cs
Compiled assembly into B.dll
$ dafny A.dfy B.dll  /spillTargetCode:1
Dafny 2.3.0.10506

Dafny program verifier finished with 1 verified, 0 errors
Compiled program written to A.cs
Errors compiling program into 
/var/folders/zg/g42b_5w97n51rgmx0jk1ysnnxxwh9d/T/du5b7dnv.0.cs(1653,17) : error CS1061: Type `B.Foo' does not contain a definition for `Bar' and no extension method `Bar' of type `B.Foo' could be found. Are you missing an assembly reference?
```

The root cause is that the Dafny definition of the B module inside B.dll (in the `DafnySourceAttribute` attribute) is included verbatim when compiling A.dfy. That means A.cs contains the Dafny-generated partial definition of B, which makes the C# compiler ignore the complete definition from B.dll and complain about the missing C# implementation.

The solution should be to include the Dafny source from DLLs when resolving references, but to not re-compile it to C#.

### PR DESCRIPTION
**Pull request by robin-aws**
Fixes #466. Involves a few different tweaks:

* Add an `IsToBeCompiled` property to `ModuleDefinition`. This is similar to the existing `IsToBeVerified` property, and is set to false by the parser when reading from DLLs.
* Allow input files to be classified as both a "Dafny source file" and a "other file" (meaning a dependency for the target platform) at the same time, so DLLs with Dafny source will still be included when compiling and/or running through the `CsharpCompiler` (and others in the future)
* Dynamically load included DLLs when running for C#, since Rosalyn can't do it automatically.

### CODE REVIEW

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 394054c6f69550682dbcc9c578c64ebfcf78f176
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=c859a1d5-d64a-40b9-a0f0-227082bb5279%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 69881384bb3422661114eca1119348d38ddc7eb5
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=80db701a-a137-4930-82f7-fef9c26bf14c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 99c48657b3e58c6e7eea13a63be3790640fdbb0a
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=dec62f78-198c-4d0b-b5ec-9cedd4d2c7de%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 9c5d61736bf34123524c102f2d7e26feeae22cf2
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=cb21853f-4098-46ca-a6a6-a38f7bdf73e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: e2f88b43548fd13fb532293fc6e32b1759447592
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b4c8b338-7035-4e20-b4d3-7ef478e649e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 7ef3add53bcd72831aa7335d20315796ef7f1db9
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b35eb8a6-6c58-4db5-8de9-855bcd398084%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: Source/Dafny/DafnyMain.cs [hunk 1]
```diff
@@ -41,13 +42,12 @@ public class DafnyFile {
...
-        Assembly.ReflectionOnlyLoad("DafnyRuntime");
-        var asm = Assembly.ReflectionOnlyLoadFrom(filePath);
+        var asm = AssemblyDefinition.ReadAssembly(filePath);
         string sourceText = null;
-        foreach (var adata in asm.GetCustomAttributesData()) {
+        foreach (var adata in asm.CustomAttributes) {
           if (adata.Constructor.DeclaringType.Name == "DafnySourceAttribute") {
             foreach (var args in adata.ConstructorArguments) {
-              if (args.ArgumentType == System.Type.GetType("System.String")) {
+              if (args.Type.Name == "String") {
```
Comment (RustanLeino) [inline Source/Dafny/DafnyMain.cs]: Does this also need to be `.FullName` and `"System.String"`, like in the other recent PR?

Comment (robin-aws) [reply Source/Dafny/DafnyMain.cs]: Yup, this PR builds on #464 and I haven't rebased since fixing that. :)

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 6e9c4be5cde1b26026ac148c18aaaef87d0a5a45
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=a3061f80-3ac3-44de-aa5f-ed513777ab9a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 6ecf2842cc8199a8a407a0ac5a684702bb969350
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=a7509e76-3d46-449c-b620-1f168e82ba80%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 2f5d1ccae8bf29ad11a32a1dc9041f5deeef4905
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=0f7b591e-1991-482d-ad00-644eacb4aa44%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**